[mkdocs] docs: add listen tx flow tutorial (#916)
NemProject/nem: mkdocs/pages/en/devbook/reference/websockets/index.md

Reviewer: ```suggestion
NEM publishes blockchain events over
[WebSockets](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API), so applications can receive live updates
without constantly polling the [REST API](../rest/nem.md).

Client applications open a WebSocket connection to any <node:> in the network and subscribe to the [channels](#channels)
they want to monitor.
When an event occurs on a channel, the node notifies every subscribed client in real time.

Some channels also accept [requests](#requests) for immediate data, similar to the REST API.
This can simplify applications that use WebSockets as their only API for both live notifications and on-demand updates.
```

In this way:
- We establish the terminology we are going to use: clients SUBSCRIBE and REQUEST, nodes NOTIFY. I don't like "push" or "send" because they work in both directions so they can be ambiguous.
- We introduce requests early on so they are not a surprise later, and we state their purpose.
- We introduce the node upfront, otherwise, you mention "the node" and it's not clear which node is that.

NEM serves WebSockets using the [STOMP](https://stomp.github.io/) messaging protocol over
[SockJS](https://github.com/sockjs/sockjs-client), on a dedicated port (`7778` by default) separate from the HTTP API
port.

> ## Session

Author: Before it only covered connection + subscription, now the full cycle with "requests" taken into account.

Reviewer: I don't understand what a "session" is.
Is it a STOMPS or SockJS concept? Or are you just describing a typical usage session?
You need to clarify this. Maybe rename the section "Typical Usage Session" ?

Author: Described session

The SockJS endpoint is `/w/messages`, for example `http://localhost:7778/w/messages`.

> A client that does not use SockJS can open a plain WebSocket to `/w/messages/websocket` directly.

Reviewer: You can't just throw the non-SockJS API here without any more information and never mention it again :D
Also, putting it in the section intro, makes it look like it is an important part of the process.

I would move all mention of the non-SockJS SPI to an info box at the end of this section AND provide more information.
So if SockJS is not used, what is the protocol? What changes? Is there anywhere the user can go to know more if they are interested in this topic?

Once the socket is open, the client drives the session with STOMP <frames:>:

1. Send a `CONNECT` frame to start the STOMP session.
2. Send a `SUBSCRIBE` frame for each <channel:> to monitor, each with a client-chosen `id`.

    !!! note "Some channels require registration"

        Some channels, like [account channels](#account-channels), deliver nothing until they receive a registration
        request.
        A client must send this request before subscribing to the channels it enables.

>         See [registration requests](#registration-requests) for the full list.

Reviewer: Instead of making this an info box, which interrupts the flow of the session, I would make it an optional step:

```suggestion
3. Send an optional [registration request](#registration-requests) to receive notifications for channels which
    do not send them automatically.
```

Also, we avoid "nothing", which is too strong a word.

    Reconnection starts a fresh session, so every channel must be subscribed again.

## Frames

> Frame

Reviewer: This glossary term is global, so we risk collisions if we use too generic words.
Since this is a pretty niche usage, I would call it "STOMP Frame" instead.

The section title should probably be called "STOMP Frames" too.

Same problem with "Channel" and "Request" below, although those are probably related to WebSocket or SockJS, not STOMP.

Author: Changed glossary entries to avoid conflict:

- STOMP Frame
- WebSocket Channel
- WebSocket Request

And H2 to:

- STOMP Frame
- Channels
- Requests

Decided not to add "Websockets" to Channels and Requests since WebSockets is already the title of the page.

> :   A plain-text message made of a command, optional `header:value` lines, and an optional body.

Reviewer: ```suggestion
:   A plain-text message adhering to the [STOMP](https://stomp.github.io/) protocol,
    made of a command, optional `header:value` lines, and an optional body.
```

If we ever use this glossary entry in any other page (doubtful 😅) the reader will appreciate the extra context.

> The client and node exchange these frames.

Reviewer: ```suggestion
The client and node exchange the following frames.
```

### `CONNECT`

> Starts the STOMP session. Sent once, right after the connection opens.

Reviewer: ```suggestion
Starts the STOMP session. The client must send this frame once, right after the connection opens.
See the [STOMP specification](https://stomp.github.io/stomp-specification-1.2.html#CONNECT_or_STOMP_Frame) for full details.
```

Make it extra clear who sends every frame (same for the rest of frames).
Include link to the right section in the spec so readers can find out about heart beats and whatever else if they want to.
Otherwise the following example is a bit puzzling.

> ```stomp

Reviewer: Add `title="Example"` or it looks like this is the mandatory message.

| ------------- | ------------------------------------------------------------------------ |
| `CONNECT`     | Starts the STOMP session. Sent once, right after the connection opens.   |
| `SUBSCRIBE`   | Subscribes to a <channel:>, with a client-chosen `id` and `destination`. |
| `SEND`        | Sends a one-time <request:> to a `/w/api` destination.                   |
| `MESSAGE`     | Delivers channel data from the node, the only frame the node sends.      |
| `UNSUBSCRIBE` | Cancels a subscription by its `id`.                                      |
| `DISCONNECT`  | Ends the session.                                                        |

## Channels

Channel
:   A destination the client subscribes to in order to receive messages the node pushes to it.

The available channels are grouped here by the type of event they report.

!!! note "Address format"

    Wherever an address appears, in a channel `destination` or a request body, it is uppercase and without hyphens.

    **Example:** `TBULEAUG2CZQISUR442HWA6UAKGWIXHDABJVIPS4`.

### Block channels

#### `/blocks`

ws:blocks
:   Notifies subscribed clients every time a new block is added to the chain.

> === "Frame"

Author: Changed from "Request body" to "Frame" so that's not confused with the new "Requests" concept.

Reviewer: Sounds good. But I'd call the tabs "Subscription frame" and "Notification body", so it's extra clear what they are.
I would even go as far as adding `:material-arrow-up-bold:` and `:material-arrow-down-bold:` (for example) at the beginning of the tab title so the direction of each message is explicit.

    Reconnection starts a fresh session, so every channel must be subscribed again.

## STOMP Frames

STOMP Frame
:   A plain-text message adhering to the [STOMP](https://stomp.github.io/) protocol,
    made of a command, optional `header:value` lines, and an optional body.

> The client and node exchange the following frames.

Reviewer: ```suggestion
The client and node exchange the following frame types.
```

### `CONNECT`

Starts the STOMP session.
The client must send this frame once, right after the connection opens.

> See the [STOMP specification](https://stomp.github.io/stomp-specification-1.2.html#CONNECT_or_STOMP_Frame) for full details.

Reviewer: I meant that you could link every frame type to its STOMP spec, not only this one :)

```stomp title="Example"
CONNECT
accept-version:1.2
heart-beat:0,0
```

### `SUBSCRIBE`

Subscribes to a <WebSocket Channel:|channel>, with a client-chosen `id` and `destination`.

```stomp title="Example"
SUBSCRIBE
id:sub-0
destination:/blocks
```

* `id` is unique only within a single connection (other clients can reuse the same value).
    It is echoed back as the `subscription` header on every message and used to `UNSUBSCRIBE` later.
* `destination` identifies the channel, so the same connection can monitor multiple channels.

### `MESSAGE`

> Delivers channel data from the node, and is the only frame the node sends.

Reviewer: ```suggestion
Delivers channel data from the node, and is the only frame type the node sends.
```

{ ... }
```

* `destination` matches the channel from the `SUBSCRIBE` frame.
* `subscription` matches the `id` from the `SUBSCRIBE` frame.
* `message-id` is a unique identifier the server assigns to each message.
* `{ ... }` is the body, a JSON object whose shape depends on the [channel](#channels).
    See the **Message body** tabs below.

### `SEND`

Sends a <request:> to a `/w/api` destination.

```stomp
SEND
destination:/w/api/account/subscribe

{ "account": "{address}" }
```

### `UNSUBSCRIBE`

Cancels a subscription by its `id`.

```stomp
UNSUBSCRIBE
id:sub-0
```

### `DISCONNECT`

Ends the session.

```stomp
DISCONNECT
```

## Channels

Channel

> :   A destination the client subscribes to in order to receive messages the node pushes to it.

Reviewer: ```suggestion
:   Node notifications are grouped into channels.
    Clients subscribe to each channel whose notifications they want to receive.
```

It seems a bit clearer to me.

> The available channels are grouped here by the type of event they report.

Reviewer: I would also mention here that all channels are subscribed via a `SUBSCRIBE` frame.
It's clear from the tables below, but this will make readers stop checking if the subscription message has changed, reducing cognitive load.

!!! note "Address format"

>     Wherever an address appears, in a channel `destination` or a request body, it is uppercase and without hyphens.

Reviewer: How did we call this type of address in the Textbook? Can you add a link?

    **Example:** `TBULEAUG2CZQISUR442HWA6UAKGWIXHDABJVIPS4`.

> ### Block channels

Reviewer: Give a short description of each category, like you did for System Channels.

```suggestion
### Block Channels
```

Also, double check headers are in Title Case.

#### `/blocks`

ws:blocks
:   Notifies subscribed clients every time a new block is added to the chain.

=== "Frame"

    ```stomp
    SUBSCRIBE
    id:sub-0
    destination:/blocks
    ```

=== "Message body"

>     [Block](../rest/nem.md#model/Block)

Reviewer: This lonely link here has always looked weird to me (even on Symbol, my bad).
Also, the section is very noisy, full of headers in code style, definition boxes, and tabs with code inside (again, my bad).

What do you think of this approach?

```markdown
<div class="centered" markdown>
<table markdown>
<tr markdown><th markdown>:material-arrow-up-bold: Subscription frame</th><th markdown>:material-arrow-down-bold: Notification frame</th></tr>
<tr markdown><td markdown>
\```stomp
SUBSCRIBE
id:sub-0
destination:/blocks/new
\```
</td><td markdown>
\```stomp
MESSAGE
destination:/blocks
subscription:sub-0
message-id:...
\```
Followed by a [Block](../rest/nem.md#model/Block) JSON body.
</td></tr></table></div>
```

I think it's much cleaner, given that we have very little text inside the table cells.

(Ignore the escaping of the code blocks, I needed it so it shows OK on GitHub)

### System channels

These channels report node status and request errors, rather than blockchain events.

#### `/node/info`

ws:node&#47;info
:   Notifies subscribed clients of the node's information, only in response to <req:w&#47;api&#47;node&#47;info>.

=== "Frame"

    ```stomp
    SUBSCRIBE
    id:sub-0
    destination:/node/info
    ```

=== "Message body"

    [Node](../rest/nem.md#model/Node)

#### `/errors`

ws:errors
:   Notifies subscribed clients when a `/w/api` <request:> fails, for example when its address payload is invalid.
    A client can subscribe to this channel right after connecting, so problems surface here instead of being silently
    dropped.

=== "Frame"

    ```stomp
    SUBSCRIBE
    id:sub-0
    destination:/errors
    ```

=== "Message body"

    ```json
    {
        "timeStamp": 67191609,
        "status": 400,
        "error": "Bad Request",
        "message": "account is not valid"
    }
    ```

## Requests

Request
:   A message the client sends to a `/w/api/...` destination to make the node act once.
    It either **registers** (a prerequisite some channels require before they deliver) or pushes a **snapshot** of

>     current data to a channel.

Reviewer: ```suggestion
Request
:   A message sent by the client to make the node deliver:
    either notifications on channels that require **registration**,
    or an immediate response containing a **snapshot** of the state of the blockchain.
```

I think it reads more clearly, but we can iterate.

For example, a request lets an application read an account's current state on demand, over the same WebSocket it uses
for live channel updates, instead of polling the [REST API](../rest/nem.md).

Every request is **read-only**.
It never changes the chain, only registers interest in receiving events or replays existing data.

> ### Registration requests

Reviewer: I would add a short intro explaining why these are needed.

Instead of "SEND frame" I would title the code blocks ":material-arrow-up-bold: Registration frame".
Maybe even use the same table layout as above (if we decide to use it), for consistency, even when it would only have one column here.

#### `/w/api/account/subscribe`

req:w&#47;api&#47;account&#47;subscribe
:   Registers the address so that [account channels](#account-channels) begin delivering events.

>     It pushes nothing back to any channel.

Reviewer: ```suggestion
req:w&#47;api&#47;account&#47;subscribe
:   Makes the node start sending notifications about the selected [account channel](#account-channels).
```

```stomp title="SEND frame"
SEND
destination:/w/api/account/subscribe

{ "account": "{address}" }
```

#### `/w/api/account/get`

req:w&#47;api&#47;account&#47;get

> :   Registers the address and pushes its current state to <ws:account&#47;{address}>.

Reviewer: Is this exactly the same as the previous one, but triggering an immediate notification?
If so, I would explain it in this way.
Always make it explicit who is sending each message and avoid using "push" on its own.

```stomp title="SEND frame"
SEND
destination:/w/api/account/get

{ "account": "{address}" }
```

> ### Snapshot requests

Reviewer: Again, short intro explaining why these are needed.

The definitions of all requests are a bit confusing. I would begin them with something like:
"This request forces the node to send an immediate notification containing..."

Instead of "SEND frame" I would title the code blocks ":material-arrow-up-bold: Request frame".
Maybe even use the same table layout as above (if we decide to use it), for consistency, even when it would only have one column here.

ws:transactions&#47;{address}
:   Notifies subscribed clients when a transaction involving the registered account is confirmed.

=== "Frame"

    ```stomp
    SUBSCRIBE
    id:sub-0
    destination:/transactions/{address}
    ```

=== "Message body"

    [TransactionMetaDataPair](../rest/nem.md#model/TransactionMetaDataPair)

> #### `/account/mosaic/owned/{address}`

Author: Discovered some more websocket channels while working on this tutorial.

=== "Frame"

    ```stomp
    SUBSCRIBE
    id:sub-0
    destination:/recenttransactions/{address}
    ```

=== "Message body"

    A list of [TransactionMetaDataPair](../rest/nem.md#model/TransactionMetaDataPair).

### System channels

These channels report node status and request errors, rather than blockchain events.

#### `/node/info`

ws:node&#47;info
:   Notifies subscribed clients of the node's information, only in response to <req:w&#47;api&#47;node&#47;info>.

=== "Frame"

    ```stomp
    SUBSCRIBE
    id:sub-0
    destination:/node/info
    ```

=== "Message body"

    [Node](../rest/nem.md#model/Node)

#### `/errors`

ws:errors
:   Notifies subscribed clients when a `/w/api` <request:> fails, for example when its address payload is invalid.
    A client can subscribe to this channel right after connecting, so problems surface here instead of being silently
    dropped.

=== "Frame"

    ```stomp
    SUBSCRIBE
    id:sub-0
    destination:/errors
    ```

=== "Message body"

    ```json
    {
        "timeStamp": 67191609,
        "status": 400,
        "error": "Bad Request",
        "message": "account is not valid"
    }
    ```

> ## Requests

Author: Added a new section to document requests properly. Originally I thought the registration only existed for account channels to work, but it turns out there are a few more that push snapshots of current data to channels.

More details in https://github.com/QuantumMechanics/nem-lightwallet/tree/master/lightwallet#nis-websocket-channels-and-data

Reviewer: Nice finding! And good idea to give it its own section.

Again, I would mention in the intro to this section (after the definition) that all requests are made via a `SEND` frame,
and the destination always begins with `/w/api/` (so you can remove this detail from the definition).

It's also worth mentioning that some of these requests do not receive any answer in return, and some others get an answer through any of the channels above.

WebSocket Request
:   A message sent by the client to make the node deliver:

>     either notifications on channels that require **registration**, or an immediate notification containing a

Author: Standardized response -> notification
